src/main/java/by/dragonsurvivalteam/dragonsurvival/client/render/entity/dragon/DragonArmorRenderLayer.java

```diff
@@ -86,6 +102,93 @@ private void renderArmorPiece(final PoseStack poseStack, final DragonEntity anim
 	}
 
+	public static ResourceLocation constructArmorTrimTexture(final Player pPlayer, EquipmentSlot pSlot) {
+		ItemStack itemstack = pPlayer.getItemBySlot(pSlot);
+		ResourceLocation existingArmorLocation = ResourceLocation.fromNamespaceAndPath(MODID, constructArmorTexture(pPlayer, pSlot));
+
+		if (itemstack.getItem() instanceof ArmorItem item) {
+			ArmorTrim trim = itemstack.get(DataComponents.TRIM);
+			if (trim != null) {
+				Optional<Resource> armorFile = Minecraft.getInstance().getResourceManager().getResource(existingArmorLocation);
+
+				String patternPath = trim.pattern().value().assetId().getPath();
+				String colorPath = trim.material().value().assetName();
+				ResourceLocation armorTrimKey = ResourceLocation.fromNamespaceAndPath(MODID, existingArmorLocation.getPath().split("\\.")[0] + "_" + patternPath + "_" + colorPath);
+
+				if (armorFile.isPresent()) {
+					try {
+						Optional<Resource> trimFile = Minecraft.getInstance().getResourceManager().getResource(ResourceLocation.fromNamespaceAndPath(MODID, "textures/armor/trims/" + patternPath + "_" + item.getType().getName() + ".png"));
+
+						if (trimFile.isPresent()) {
+							if ((Minecraft.getInstance().getTextureManager().getTexture(armorTrimKey) instanceof DynamicTexture)) {
+								return armorTrimKey;
+							}
+							NativeImage image = new NativeImage(512, 512, true);
+
+							InputStream textureStream = armorFile.get().open();
+							NativeImage armorImage = NativeImage.read(textureStream);
+							textureStream.close();
+
+							textureStream = trimFile.get().open();
+							NativeImage trimImage = NativeImage.read(textureStream);
+							textureStream.close();
+
+							TextColor tc = trim.material().value().description().getStyle().getColor();
+							// Not the most elegant solution, but the best way I could find to get a single color reliably...
+							// TODO: something better
+							if (tc != null) {
+								float[] trimBaseHSB = new float[3];
+								float[] armorHSB = new float[3];
+								float[] trimHSB = new float[3];
+								Color trimBaseColor = new Color(tc.getValue());
+								Color.RGBtoHSB(trimBaseColor.getBlue(), trimBaseColor.getGreen(), trimBaseColor.getRed(), trimBaseHSB);
+
+								for (int x = 0; x < armorImage.getWidth(); x++) {
+									for (int y = 0; y < armorImage.getHeight(); y++) {
+										Color armorColor = new Color(armorImage.getPixelRGBA(x, y), true);
+										Color trimColor = new Color(trimImage.getPixelRGBA(x, y), true);
+										Color.RGBtoHSB(armorColor.getRed(), armorColor.getGreen(), armorColor.getBlue(), armorHSB);
+										Color.RGBtoHSB(trimColor.getRed(), trimColor.getGreen(), trimColor.getBlue(), trimHSB);
+
+										if (trimColor.getAlpha() != 0) {
+											image.setPixelRGBA(x, y, Color.HSBtoRGB(trimBaseHSB[0], trimBaseHSB[1], trimHSB[2]));
```

Comment:
is there a reason why two values are from `trimBase` and one is from `trim`?

Reply:
It takes the hue from the armor trim color, and the brightness and contrast from the armor trim pixel.  The armor trim file can be grayscale.

Reply:
but if the order is hue, saturation and brightness
doesn't it take hue and saturation from one file and brightness from another
instead of hue from one and brightness + saturation from another

Reply:
so is this correct or not?

Reply:
#611